cluster-autoscaler/main.go

```diff
@@ -0,0 +1,227 @@
+)
+
+func createAutoscalerOptions() core.AutoscalerOptions {
+	autoscalingOpts := core.AutoscalingOptions{
+		CloudConfig: *cloudConfig,
+		CloudProviderName: *cloudProviderFlag,
+		MaxTotalUnreadyPercentage: *maxTotalUnreadyPercentage,
+		OkTotalUnreadyCount: *okTotalUnreadyCount,
+		EstimatorName: *estimatorFlag,
+		ExpanderName: *expanderFlag,
+		MaxEmptyBulkDelete: *maxEmptyBulkDeleteFlag,
+		MaxGratefulTerminationSec: *maxGratefulTerminationFlag,
+		MaxNodeProvisionTime: *maxNodeProvisionTime,
+		MaxNodesTotal: *maxNodesTotal,
+		UnregisteredNodeRemovalTime: *unregisteredNodeRemovalTime,
+		ScaleDownDelay: *scaleDownDelay,
+		ScaleDownEnabled: *scaleDownEnabled,
+		ScaleDownTrialInterval: *scaleDownTrialInterval,
+		ScaleDownUnneededTime: *scaleDownUnneededTime,
+		ScaleDownUnreadyTime: *scaleDownUnreadyTime,
+		ScaleDownUtilizationThreshold: *scaleDownUtilizationThreshold,
+		VerifyUnschedulablePods: *verifyUnschedulablePods,
+	}
+
+	configFetcherOpts := dynamic.ConfigFetcherOptions{
+		ConfigMapName: *configMapName,
+		Namespace: *namespace,
+	}
+
+	return core.AutoscalerOptions{
+		AutoscalingOptions: autoscalingOpts,
+		ConfigFetcherOptions: configFetcherOpts,
+	}
+}
+
+func createKubeClient() kube_client.Interface {
+	url, err := url.Parse(*kubernetes)
+	if err != nil {
+		glog.Fatalf("Failed to parse Kuberentes url: %v", err)
+	}
+
+	kubeConfig, err := config.GetKubeClientConfig(url)
+	if err != nil {
+		glog.Fatalf("Failed to build Kuberentes client configuration: %v", err)
+	}
+
+	return kube_client.NewForConfigOrDie(kubeConfig)
+}
+
+// In order to meet interface criteria for LeaderElectionConfig we need to
+// take stop channell as an argument. However, since we are committing a suicide
+// after loosing mastership we can safely ignore it.
+func run(_ <-chan struct{}) {
+	kubeClient := createKubeClient()
+	kubeEventRecorder := kube_util.CreateEventRecorder(kubeClient)
+	opts := createAutoscalerOptions()
+	autoscaler := core.NewAutoscaler(opts, kubeClient, kubeEventRecorder)
+
+	for {
+		select {
+		case <-time.After(*scanInterval):
+			{
+				loopStart := time.Now()
+				metrics.UpdateLastTime("main")
+
+				autoscaler.RunOnce(loopStart)
+
+				metrics.UpdateDuration("main", loopStart)
+			}
+		}
+	}
+}
+
+func main() {
+	leaderElection := kube_leaderelection.DefaultLeaderElectionConfiguration()
+	leaderElection.LeaderElect = true
+
+	kube_leaderelection.BindFlags(&leaderElection, pflag.CommandLine)
+	flag.Var(&nodeGroupsFlag, "nodes", "sets min,max size and other configuration data for a node group in a format accepted by cloud provider."+
+		"Can be used multiple times. Format: <min>:<max>:<other...>")
+	kube_flag.InitFlags()
+
+	glog.Infof("Cluster Autoscaler %s", ClusterAutoscalerVersion)
+
+	correctEstimator := false
+	for _, availableEstimator := range estimator.AvailableEstimators {
+		if *estimatorFlag == availableEstimator {
+			correctEstimator = true
+		}
+	}
```

Comment:
I tempted to move this validation logic of the estimatorFlag into `estimator` package, probably as something like `func IsValidName(estimatorName string) bool`, but kept untouched thinking doing so would be out of scope of this pull request.